### PR TITLE
Adds ability for the user to choose if they want to double encode entities within textareas or not when using ActiveForm. Applies enhancement from #12988.

### PR DESCRIPTION
**Pull request by cyphix333**
| Q             | A
| ------------- | ---
| Is bugfix?    | no
| New feature?  | yes
| Breaks BC?    | no
| Tests pass?   | yes
| Fixed issues  | #12988

### CODE REVIEW

**Review by samdark — CHANGES_REQUESTED** [review]
Need a changelog and a unit test.

File: framework/helpers/BaseHtml.php [hunk 1]
```diff
@@ -660,12 +660,21 @@ public static function fileInput($name, $value = null, $options = [])
...
+     * - `doubleEncode`: whether to encode HTML entities in `$value`. If false, HTML entities in `$value` will not
+     * be further encoded.
+     *
      * @return string the generated text area tag
      */
     public static function textarea($name, $value = '', $options = [])
     {
         $options['name'] = $name;
-        return static::tag('textarea', static::encode($value), $options);
+        $doubleEncode = (isset($options['doubleEncode'])) ? $options['doubleEncode'] : true;
```
Comment (SilverFire) [inline framework/helpers/BaseHtml.php]: All these lines could be replaced with one:

```php
ArrayHelper::remove($options, 'doubleEncode', true);
```

**Review by SilverFire — CHANGES_REQUESTED** [review]

File: framework/helpers/BaseHtml.php (hunk 1 above)
Comment (cyphix333) [inline framework/helpers/BaseHtml.php]: Ahh nice, haven't used `ArrayHelper` enough to know you could do that. I have made the changes locally.

File: framework/helpers/BaseHtml.php [hunk 2]
```diff
@@ -660,12 +660,21 @@ public static function fileInput($name, $value = null, $options = [])
      * the attributes of the resulting tag. The values will be HTML-encoded using [[encode()]].
      * If a value is null, the corresponding attribute will not be rendered.
      * See [[renderTagAttributes()]] for details on how attributes are being rendered.
+     * The following special options are recognized:
+     *
+     * - `doubleEncode`: whether to encode HTML entities in `$value`. If false, HTML entities in `$value` will not
```
Comment (rob006) [inline framework/helpers/BaseHtml.php]: `false` should be wrapped by `` ` ``

**Comment by cyphix333** [comment]
Sorry for the delay in getting the revised version up but I am having some problems getting phpunit working, will finalize this as soon as I solve the issues.

**Comment by cyphix333** [comment]
Ran a unit test on the `helpers` group and there was a failure that was unrelated to the changes I did, line `958` of `helpers/HtmlTest.php`.

File: framework/helpers/BaseHtml.php [hunk 3]
```diff
@@ -660,12 +660,20 @@ public static function fileInput($name, $value = null, $options = [])
...
+     * - `doubleEncode`: whether to encode HTML entities in `$value`. If `false`, HTML entities in `$value` will not
+     * be further encoded.
+     *
      * @return string the generated text area tag
      */
     public static function textarea($name, $value = '', $options = [])
     {
         $options['name'] = $name;
-        return static::tag('textarea', static::encode($value), $options);
+        // Get value of `doubleEncode` option and then remove it from the array as it is not
```
Comment (dynasource) [inline framework/helpers/BaseHtml.php]: can this be done in 1 line to make it less obtrusive?

File: framework/helpers/BaseHtml.php [hunk 4]
```diff
@@ -660,12 +660,20 @@ public static function fileInput($name, $value = null, $options = [])
      * the attributes of the resulting tag. The values will be HTML-encoded using [[encode()]].
      * If a value is null, the corresponding attribute will not be rendered.
      * See [[renderTagAttributes()]] for details on how attributes are being rendered.
+     * The following special options are recognized:
+     *
+     * - `doubleEncode`: whether to encode HTML entities in `$value`. If `false`, HTML entities in `$value` will not
```
Comment (dynasource) [inline framework/helpers/BaseHtml.php]: the naming `doubleEncode` does not align with the definition.

**Review by dynasource — CHANGES_REQUESTED** [review]
need unit tests

File: framework/helpers/BaseHtml.php (hunk 4 above)
Comment (cyphix333) [inline framework/helpers/BaseHtml.php]: How so? It's actually a copy straight from the `encode` method and seems to say what it does perfectly.

File: framework/helpers/BaseHtml.php (hunk 3 above)
Comment (cyphix333) [inline framework/helpers/BaseHtml.php]: If you're referencing the comment then I think more _readable_ code/comments is far better than something slightly obtrusive.

File: framework/helpers/BaseHtml.php (hunk 4 above)
Comment (dynasource) [inline framework/helpers/BaseHtml.php]: the word 'double' is not explained

File: framework/helpers/BaseHtml.php (hunk 3 above)
Comment (dynasource) [inline framework/helpers/BaseHtml.php]: if you can't explain it in 1 line, you don't understand it well enough.

File: framework/helpers/BaseHtml.php (hunk 4 above)
Comment (cyphix333) [inline framework/helpers/BaseHtml.php]: Does it need to be as it references `HTML entities` which means they are already encoded.

File: framework/helpers/BaseHtml.php (hunk 3 above)
Comment (cyphix333) [inline framework/helpers/BaseHtml.php]: Well feel free to re-word it as you like, but I understand it fine :)

File: framework/helpers/BaseHtml.php (hunk 4 above)
Comment (dynasource) [inline framework/helpers/BaseHtml.php]: of course. You are using the word 'double' so people are looking for that word in the definition. Why is it `doubleEncode` and not just named `encode`

File: framework/helpers/BaseHtml.php (hunk 3 above)
Comment (dynasource) [inline framework/helpers/BaseHtml.php]: could be more declarative instead of imperative

File: framework/helpers/BaseHtml.php (hunk 4 above)
Comment (dynasource) [inline framework/helpers/BaseHtml.php]: you don't have to answer that to me ;) It's what developers are questioning if they see the parameter 'doubleEncode'

File: framework/helpers/BaseHtml.php (hunk 3 above)
Comment (cyphix333) [inline framework/helpers/BaseHtml.php]: Updated - hopefully this is to your liking better.

File: framework/helpers/BaseHtml.php (hunk 4 above)
Comment (cyphix333) [inline framework/helpers/BaseHtml.php]: I was just following the standard way it is done. :)

**Review by dynasource — APPROVED** [review]

**Review by dynasource — CHANGES_REQUESTED** [review]
don't forget about the unit tests

**Comment by cyphix333** [comment]
@dynasource I have already done unit tests, please see my [previous post](https://github.com/yiisoft/yii2/pull/12994#issuecomment-261773453).

**Comment by dynasource** [comment]
you have to add them to cover your usecase (so we don't loose you solution in future commits)

**Comment by cyphix333** [comment]
@dynasource Sorry I haven't used phpunit before and was unsure what you wanted. I have attached the output of the test.

[unit-test.txt](https://github.com/yiisoft/yii2/files/606065/unit-test.txt)

**Comment by cyphix333** [comment]
@dynasource Did you need anything else from me?

**Comment by dynasource** [comment]
@cyphix333, Brett, we have to have unit tests (like https://github.com/yiisoft/yii2/pull/13079/files#diff-e4de87b687b5c876d6209e73ed20863eR97) in order for PR's to accept (and to stay stable in the future)

File: framework/helpers/BaseHtml.php [hunk 5]
```diff
@@ -660,12 +660,19 @@ public static function fileInput($name, $value = null, $options = [])
      * the attributes of the resulting tag. The values will be HTML-encoded using [[encode()]].
      * If a value is null, the corresponding attribute will not be rendered.
      * See [[renderTagAttributes()]] for details on how attributes are being rendered.
+     * The following special options are recognized:
+     *
+     * - `doubleEncode`: whether to double encode HTML entities in `$value`. If `false`, HTML entities in `$value` will not
+     * be further encoded.
```
Comment (cebe) [inline framework/helpers/BaseHtml.php]: Please add `This option is available since version 2.0.11.`

**Review by cebe — APPROVED** [review]

**Comment by SilverFire** [comment]
Merged with minor adjustments: 7f46ab33
Thank you

**Pull request closed without merging**